### ISSUE TITLE
[Performance] Disable PropTypes validation in production builds

### ISSUE THREAD

**Issue by rigdern** [facebook/react#6864]
React can run faster in production builds if it doesn't check PropTypes. Looking at the code, I don't see any `__DEV__` checks which would prevent PropTypes validation from running in production builds:
  - [createElement](https://github.com/facebook/react/blob/db6ac5c01c4ad669db7ca264bc81ae5b3d6dfa01/src/isomorphic/classic/element/ReactElementValidator.js#L217)
  - [cloneElement](https://github.com/facebook/react/blob/db6ac5c01c4ad669db7ca264bc81ae5b3d6dfa01/src/isomorphic/classic/element/ReactElementValidator.js#L262)

We're seeing functions like `validatePropTypes` and `checkType` pop up while doing performance analysis on our app.

**Comment by jimfb** [comment]
You're seeing validatePropTypes and checkType in the production build?  :/.

The production build is minified.  How would you even find those functions?

**Comment by jimfb** [comment]
Yeah, I'm pretty sure the functions you linked are not called in production.

**Comment by rigdern** [comment]
@jimfb What prevents them from being called in production builds?

**Comment by gaearon** [comment]
You are looking at ReactElementValidator. Its functions are not used in production build. If you look for the matches in the source, you will see that it's only used in DEV.

**Comment by rigdern** [comment]
Thanks for looking into this and providing an explanation.

It looks like we made a mistake when we were verifying our theory. We modified React and deleted `validatePropTypes` from the build to determine whether it made things noticeably faster. However, when we produced a bundle of our app, we accidentally included a minified version of React's dev build instead of production build. Sorry for the trouble.
